### PR TITLE
=per #19828 pop the internal stash when the writing was finished in t…

### PR DESCRIPTION
**Pull request by MQ-EL**
…he processing commands state

### CODE REVIEW

File: project/MiMa.scala [hunk 1]
```diff
@@ -607,7 +607,10 @@ object MiMa extends AutoPlugin {
 
         // #19440
         ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.PipeToSupport.pipeCompletionStage"),
-        ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.FutureTimeoutSupport.afterCompletionStage")
+        ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.FutureTimeoutSupport.afterCompletionStage"),
+        
+        //#19828
+        ProblemFilters.exclude[AbstractMethodProblem]("akka.persistence.Eventsourced#ProcessingState.onWriteMessageComplete")
```
Comment (MQ-EL) [inline project/MiMa.scala]: is it reasonable?

**Comment by MQ-EL** [comment]
Refs #19828

**Comment by ktoso** [comment]
OK TO TEST

File: project/MiMa.scala (hunk 1 above)
Comment (johanandren) [inline project/MiMa.scala]: `ProcessingState` is private so I think this is safe.

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala [hunk 2]
```diff
@@ -183,7 +186,7 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
   /** INTERNAL API. */
   override protected[akka] def aroundPreRestart(reason: Throwable, message: Option[Any]): Unit = {
     try {
-      internalStash.unstashAll()
+      unstashInternally(all = true)
```
Comment (johanandren) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: I think these could just as well have been left `internalStash.unstashAll`

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala [hunk 3]
```diff
@@ -538,6 +541,8 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
    * Common receive handler for processingCommands and persistingEvents
    */
   private abstract class ProcessingState extends State {
+    override def recoveryRunning: Boolean = false
+
```
Comment (johanandren) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: Why was this change needed? It's not referenced anywhere in the other changes as far as I can see.

Comment (johanandren) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: Ah, I see, it was in both the concrete states.

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala [hunk 4]
```diff
@@ -604,12 +608,13 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
...
-        internalStash.unstashAll()
-      else
-        internalStash.unstash()
+      if (pendingStashingPersistInvocations > 0) changeState(persistingEvents)
+      else unstashInternally(all = err)
+    }
+
+    override def onWriteMessageComplete(err: Boolean): Unit = {
+      pendingInvocations.pop()
+      unstashInternally(all = err)
```
Comment (johanandren) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: So, this is the actual bugfix, right, the rest is just generic cleanup?

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala [hunk 5]
```diff
@@ -108,9 +114,13 @@ abstract class PersistentActorStashingSpec(config: Config) extends PersistenceSp
       persistentActor ! Cmd("a")
       persistentActor ! Cmd("b")
       persistentActor ! Cmd("c")
+      persistentActor ! Cmd("d")
+      persistentActor ! Cmd("e")
```
Comment (johanandren) [inline akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala]: A separate test covering this bug would be nicer than appending it to existing test cases.

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala (hunk 2 above)
Comment (MQ-EL) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: If the ``internalStash.unstash/all`` method has been encapsulated, perhaps call it directly in other places would not be a good idea for me.

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala (hunk 4 above)
Comment (MQ-EL) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: yes, it is.

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala (hunk 2 above)
Comment (johanandren) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: I wouldn't think of it as `unstashAll` being encapsulated but the case of conditional `unstashAll` which this isn't. Not important though, if you feel strongly about this, let it be as is.

Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: I agree with @johanandren, I don't see the benefit of that method

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala [hunk 6]
```diff
@@ -604,12 +608,13 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
...
       if (eventBatch.nonEmpty) flushBatch()
 
-      if (pendingStashingPersistInvocations > 0)
-        changeState(persistingEvents)
-      else if (err)
-        internalStash.unstashAll()
-      else
-        internalStash.unstash()
+      if (pendingStashingPersistInvocations > 0) changeState(persistingEvents)
+      else unstashInternally(all = err)
```
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: ok, here is the benefit, alright to keep then

File: project/MiMa.scala (hunk 1 above)
Comment (patriknw) [inline project/MiMa.scala]: diff of javap of EventSouced looks good

**Comment by patriknw** [comment]
LGTM

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala (hunk 5 above)
Comment (drewhk) [inline akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala]: +1

**Comment by drewhk** [comment]
LGTM, I would also appreciate a new test case though.

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala (hunk 2 above)
Comment (MQ-EL) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: Got it, thanks @johanandren @patriknw :)

**Comment by MQ-EL** [comment]
BTW: would [contrib-sharding-config](https://github.com/akka/akka/blob/master/akka-contrib/src/main/resources/reference.conf#L10) need to remove?

**Comment by patriknw** [comment]
good catch about contrib-sharding-config, @MQ-EL 
it's indeed obsolite

**Comment by patriknw** [comment]
PLS REBUILD

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala [hunk 7]
```diff
@@ -112,9 +153,11 @@ abstract class PersistentActorStashingSpec(config: Config) extends PersistenceSp
       expectMsg("c")
       expectMsg("a")
     }
+  }
 
+  def stashWithServralMessages[T <: NamedPersistentActor : ClassTag](): Unit = {
```
Comment (johanandren) [inline akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala]: Typo, should be "Several"

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala [hunk 8]
```diff
@@ -134,18 +179,29 @@ abstract class PersistentActorStashingSpec(config: Config) extends PersistenceSp
...
+    behave like stashWithServralMessages[UserStashManyPersistentActor]()
+    behave like stashUnderFailures[UserStashFailurePersistentActor]()
+  }
+  
+  "Stashing(unstashAll called in handler) in a persistent actor" must {
+    behave like stash[UserStashWithinHandlerPersistentActor]()
+    behave like stashWithServralMessages[UserStashWithinHandlerManyPersistentActor]()
+    behave like stashUnderFailures[UserStashWithinHandlerFailureCallbackPersistentActor]()
+  }
+
```
Comment (johanandren) [inline akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala]: Thanks! :+1:

**Comment by johanandren** [comment]
LGTM

**Comment by MQ-EL** [comment]
In my local,  it is ok. what does it happened?

**Comment by ktoso** [comment]
Binary compat check failed: 

```
[info] akka-persistence: found 1 potential binary incompatibilities while checking against com.typesafe.akka:akka-persistence_2.11:2.4.0  (filtered 14)
[error]  * in current version there is abstract method onWriteMessageComplete(Boolean)Unit in class akka.persistence.Eventsourced#ProcessingState, which does not have a correspondent
[error]    filter with: ProblemFilters.exclude[ReversedAbstractMethodProblem]("akka.persistence.Eventsourced#ProcessingState.onWriteMessageComplete")
```

Since `Eventsourced` is private API, we can add the exception AFAICS.

**Comment by ktoso** [comment]
You can run the checl locally via `mimaReportBinaryIssues` or `validatePullRequest` (also runs tests, only in modules which given PR could affect)

**Comment by patriknw** [comment]
@MQ-EL what is missing here? is it only the mima filter? can you fix that, or would you like me to do it?

**Comment by MQ-EL** [comment]
Sorry for late, did i miss something? I have already add mima filter ,but it seems not effect. I will try fix it today.

**Comment by MQ-EL** [comment]
Hi @patriknw , I found that the version of this pr branch is conflict with master( ```MiMa.scala``` && mima version from 2.1.7 to 2.1.9, which ```abstractMethod``` changed). How  to surround it? recreate pr or you do?

**Comment by patriknw** [comment]
@MQ-EL can't you rebase it? I'm sorry for the inconveniences. I can do that for you if you like?

**Comment by MQ-EL** [comment]
ok, I will rebase it, but wait for tomorrow morning. sorry for that

**Comment by drewhk** [comment]
@patriknw this is mergeable right?

**Pull request merged**